src/serial_io.rs

```diff
@@ -0,0 +1,45 @@
+use heapless::spsc::Queue;
+use esp32c3_hal::{ 
+    Serial,
+    pac::UART0,
+};
+use esp_hal_common::{ 
+    serial::Instance, 
+    interrupt,
+    interrupt::CpuInterrupt::*,
+    Cpu::*,
+};
+use crate::protocol::InputIO;
+use crate::targets::esp32c3 as target;
+
+const RX_QUEUE_SIZE: usize = target::MAX_WRITE_BLOCK + 0x400;
+static mut RX_QUEUE: Queue<u8, RX_QUEUE_SIZE> = Queue::new();
+
+impl<'a, T: Instance> InputIO for Serial<T> {
+    fn recv(&mut self) -> u8 {
+        let mut consumer = unsafe { RX_QUEUE.split().1 };
+
+        unsafe{
+            while consumer.ready() == false { }
```

Comment:
Unsafe is unnecessary here.

```rust
loop {
    if let Some(byte) = consumer.dequeue() {
      return byte
    }
}
```
This will do the same thing. `dequeue` already does the get_unchecked optimization inside it, see: https://docs.rs/heapless/latest/src/heapless/spsc.rs.html#261